src/fides/api/service/storage/streaming/smart_open_streaming_storage.py

```diff
@@ -0,0 +1,999 @@
+        privacy_request: PrivacyRequest,
+        buffer_config: StreamingBufferConfig,
+        batch_size: int,
+    ) -> Optional[AnyHttpUrlString]:
+        """Handle HTML format uploads with DSR report generation.
+
+        Args:
+            config: Upload configuration
+            data: Data to upload
+            privacy_request: Privacy request object
+            buffer_config: Buffer configuration
+            batch_size: Number of attachments to process in each batch
+
+        Returns:
+            presigned_url or None if URL generation fails
+        """
+        # Generate the DSR report first
+        try:
+            dsr_buffer = DsrReportBuilder(
+                privacy_request=privacy_request,
+                dsr_data=data,
+            ).generate()
+            # Reset buffer position to ensure it can be read multiple times
+            dsr_buffer.seek(0)
+        except Exception as e:
+            logger.error(f"Failed to generate DSR report: {e}")
+            raise StorageUploadError(f"Failed to generate DSR report: {e}") from e
+
+        # Check if there are attachments to include
+        all_attachments = self._collect_and_validate_attachments(data)
+
+        if not all_attachments:
+            # No attachments, just upload the DSR report
+            stream_dsr_buffer_to_storage(
+                self.storage_client,
+                config.bucket_name,
+                config.file_key,
+                dsr_buffer,
+            )
+
+            try:
+                return self.storage_client.generate_presigned_url(
+                    config.bucket_name, config.file_key
+                )
+            except Exception as e:
+                logger.error(
+                    f"Failed to generate presigned URL for {config.bucket_name}/{config.file_key}: {e}"
+                )
+                raise StorageUploadError(
+                    f"Failed to generate presigned URL: {e}"
+                ) from e
+        logger.debug(
+            f"Creating HTML DSR report ZIP with {len(all_attachments)} attachments"
+        )
+
+        # Create ZIP generator with DSR report files
+        dsr_files_generator = create_dsr_report_files_generator(
+            dsr_buffer,
+            all_attachments,
+            config.bucket_name,
+            config.max_workers,
+            batch_size,
+        )
+
+        # Create ZIP generator with attachment files
+        attachment_files_generator = self._create_attachment_files(all_attachments)
+
+        # Combine both generators and stream the complete ZIP to storage
+        with self.storage_client.stream_upload(
+            config.bucket_name,
+            config.file_key,
+            content_type="application/zip",
+        ) as upload_stream:
+            for chunk in stream_zip(attachment_files_generator):
+                upload_stream.write(chunk)
+
+            for chunk in stream_zip(dsr_files_generator):
+                upload_stream.write(chunk)
```

Comment:
Each invocation of `stream_zip` will create a separate zip file, so calling this twice will result in two concatenated ZIP files being written to the upload stream. Apparently some clients might treat this as an invalid archive or only read the last ZIP file. It's preferred to `chain` the streams and write them out with a single call to `stream_zip`.
```suggestion
        from itertools import chain

        combined_entries = chain(attachment_files_generator, dsr_files_generator)
        with self.storage_client.stream_upload(
            config.bucket_name,
            config.file_key,
            content_type="application/zip",
        ) as upload_stream:
            for chunk in stream_zip(combined_entries):
                upload_stream.write(chunk)
```